packages/firebase_auth/lib/firebase_auth.dart

```diff
@@ -158,6 +168,32 @@ class FirebaseAuth {
     return currentUser;
   }
 
+  Future<FirebaseUser> signInWithCustomToken({@required String token}) async {
+    assert(token != null);
+    final Map<String, dynamic> data = await channel.invokeMethod(
+      'signInWithCustomToken',
+      <String, String>{
+        'token': token,
+      },
+    );
+    final FirebaseUser currentUser = new FirebaseUser._(data);
+    return currentUser;
+  }
+
+  /// Start receiving changes about AuthState
+  /// See [onAuthStateChanged]
+  Future<Null> startListeningAuthState() {
```

Comment:
Instead of exposing public methods for starting and stopping listening, you could start and stop listening in the `onListen` and `onCancel` handlers of the `StreamController`. This ensures that one client that is trying to clean up after itself won't inadvertently trample over another client that is still listening for auth state changed events.